File: lib/ruby/015_solution.rb
# The following iterative sequence is defined for the set of positive integers:

# n → n/2 (n is even)
# n → 3n + 1 (n is odd)

# Using the rule above and starting with 13, we generate the following sequence:

# 13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
# It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

# Which starting number, under one million, produces the longest chain?
class Collatz
  attr_accessor :length

  def initialize(starting_number)  
    @collatz_sequence_holder = []
    @last_number = starting_number
    @collatz_sequence_holder << @last_number
    collatz_sequence
  end
  
  def collatz_sequence
    while @last_number != 1
      @last_number = find_next_number
      @collatz_sequence_holder << @last_number
    end
  end

  def find_next_number
    if @last_number % 2 == 0
      next_even
    elsif @last_number % 2 != 0
      next_odd
    end
  end

  def next_even
    @last_number / 2
  end

  def next_odd
    (3 * @last_number) + 1
  end

  def length
    @collatz_sequence_holder.length
  end

end



def find_longest
  longest_sequence = {0 => 0}
  sequence = 0
  999999.downto(2) do |number|
    sequence = Collatz.new(number).length
    if sequence > longest_sequence.values[0]
      longest_sequence = {number => sequence}
    end
  end
  puts "the longest number is #{longest_sequence.keys[0]} and the length is #{longest_sequence.values[0]}"
end

find_longest

File: lib/javascript/015_solution.js
// The following iterative sequence is defined for the set of positive integers:

// n → n/2 (n is even)
// n → 3n + 1 (n is odd)

// Using the rule above and starting with 13, we generate the following sequence:

// 13 → 40 → 20 → 10 → 5 → 16 → 8 → 4 → 2 → 1
// It can be seen that this sequence (starting at 13 and finishing at 1) contains 10 terms. Although it has not been proved yet (Collatz Problem), it is thought that all starting numbers finish at 1.

// Which starting number, under one million, produces the longest chain?

function LongestCollatz(){
  this.answer = 0
  this.currentNumber = 0
  this.longestSequenze = 0
  this.longestStartingInteger = 0 
}

LongestCollatz.prototype.nextNumber = function(){
  if(this.currentNumber % 2 === 0){
    this.nextEven();
  }
  else{
    this.nextOdd();
  }
}

LongestCollatz.prototype.nextEven= function(){
  this.currentNumber = this.currentNumber / 2;
}

LongestCollatz.prototype.nextOdd = function(){
  this.currentNumber = (3 * this.currentNumber) + 1;
}

LongestCollatz.prototype.runner = function(){
  for(var i = 2; i < 1000000; i++){
  var sequenceHolder = [i]
  this.currentNumber = i
  while(this.currentNumber !== 1){
    this.nextNumber();
    sequenceHolder.push(this.currentNumber);
  }
  if(sequenceHolder.length > this.longestSequenze){
    this.longestSequenze = sequenceHolder.length
    this.longestStartingInteger = i
  }

  }
  console.log(this.longestStartingInteger);
}


var l = new LongestCollatz
l.runner();